shared/src/main/java/eu/maveniverse/maven/toolbox/shared/internal/PomTransformerSink.java

```diff
@@ -0,0 +1,143 @@
+
+    /**
+     * Creates a directory sink.
+     *
+     * @param output The output.
+     * @param pom The POM path, if not existing, will be created (as "blank").
+     * @param blankPomSupplier Required, if pom path points to a non-existent POM file.
+     * @param artifactMatcher The artifact matcher.
+     * @param artifactMapper The artifact mapper.
+     * @param transformations The transformations to apply.
+     * @throws IOException In case of IO problem.
+     */
+    private PomTransformerSink(
+            Output output,
+            Path pom,
+            Supplier<String> blankPomSupplier,
+            Predicate<Artifact> artifactMatcher,
+            Function<Artifact, Artifact> artifactMapper,
+            Function<Artifact, PomTransformer.Transformation> transformations)
+            throws IOException {
+        this.output = requireNonNull(output, "output");
+        this.pom = requireNonNull(pom, "pom").toAbsolutePath();
+        if (Files.isRegularFile(pom)) {
+            existingPom = true;
+            Files.copy(
+                    pom,
+                    pom.getParent().resolve(getPomPath().getFileName() + ".bak"),
+                    StandardCopyOption.REPLACE_EXISTING);
+        } else {
+            Files.createDirectories(pom.getParent());
+            Files.writeString(pom, blankPomSupplier.get(), StandardCharsets.UTF_8);
+            existingPom = false;
+        }
+        this.artifactMatcher = requireNonNull(artifactMatcher, "artifactMatcher");
+        this.artifactMapper = requireNonNull(artifactMapper, "artifactMapper");
+        this.transformations = requireNonNull(transformations, "transformations");
+
+        this.applicableTransformations = new ArrayList<>();
+    }
+
+    public Path getPomPath() {
+        return pom;
+    }
+
+    @Override
+    public void accept(Artifact artifact) throws IOException {
+        requireNonNull(artifact, "artifact");
+        if (artifactMatcher.test(artifact)) {
+            PomTransformer.Transformation transformation = transformations.apply(artifactMapper.apply(artifact));
+            if (transformation != null) {
+                output.chatter("Accepted {}", artifact);
+                applicableTransformations.add(transformation);
+            }
+        }
+    }
+
+    @Override
+    public void cleanup(Exception e) {
+        try {
+            Files.deleteIfExists(pom);
+            if (existingPom) {
+                // return existing one backup
+                Files.move(
+                        pom.getParent().resolve(getPomPath().getFileName() + ".bak"),
+                        pom,
+                        StandardCopyOption.REPLACE_EXISTING);
+            }
+        } catch (IOException ex) {
+            // ignore
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        new PomTransformer(pom, StandardCharsets.UTF_8, PomTransformer.SimpleElementWhitespace.EMPTY)
```

Comment:
`SimpleElementWhitespace.AUTODETECT_PREFER_EMPTY` might be more respectful to folks having other preferences.